website/docs/02-separation-logic-and-biabduction.md

```diff
@@ -0,0 +1,264 @@
+
+This is an instance of a general pattern.
+There is a rule that lets you go from smaller to bigger specifications
+
+---
+
+<Math code="%5Cfrac%7B%5Clbrace%20pre%20%5Crbrace%20prog%20%5Clbrace%20post%20%5Crbrace%7D%7B%5Clbrace%20pre%20*%20frame%20%5Crbrace%20prog%20%5Clbrace%20post%20*%20frame%20%5Crbrace%7D" />
+
+---
+
+Our passage from (spec) to (use) is obtained by taking
+
+- <Math code="pre" /> to be <Math code="r%5C_1%20%5Cmapsto%20open" />
+- <Math code="post" /> to be <Math code="r%5C_1%20%5Cmapsto%20closed" />, and
+- <Math code="frame" /> to be <Math code="r%5C_2%20%5Cmapsto%20open" />
+
+This rule is called the _frame rule_ of separation logic. It is named after the frame problem, a classic problem in artificial intelligence.
+Generally, the <Math code="frame" /> describes state that remains unchanged; the terminology comes from the analogy of
+a background scene in an animation as unchanging while the objects and characters within the scene change.
+
+The frame rule is the key to the principle of local reasoning in separation logic: reasoning and specifications
+should concentrate on the resources that a program accesses (the footprint), without mentioning what
+doesn't change.
+
+## Bi-abduction
+
+Bi-abduction is a form of logical inference for separation logic which automates the key ideas about local
+reasoning.
+
+Usually, logic works with validity or entailment statements like
+
+---
+
+<Math code="A%20%5Cvdash%20B" />
+
+---
+
+which says that <Math code="A" /> implies <Math code="B" />. Infer uses an extension of this inference question in an internal
+theorem prover while it runs over program statements.
+Infer's question
+
+---
+
+<Math code="A%20*%20%3Fantiframe%20%5Cvdash%20B%20*%20%3Fframe" />
+
+---
+
+is called _bi-abduction_. The problem here is for the theorem prover to <i> discover </i> a pair of frame and antiframe formulae that make the entailment statement valid.
+
+Global analyses of large programs are normally computational untractable. However,
+bi-abduction allows to break the large analysis of a large program in small independent analyses of its procedures. This gives Infer the ability to scale independently of the size of the analyzed code. Moreover, by breaking the analysis in small
+independent parts, when the full program is analyzed again because
+of a code change the analysis results of the unchanged part of the
+code can be reused and only the code change needs to be re-analyzed. This process is called incremental analysis and it
+is very powerful when integrating a static analysis tool like infer in a development environment.
+
+In order to be able to decompose a global analysis in small independent analyses, let's first consider how a function
+call is analyzed in separation logic. Assume we have the following spec for a function <Math code="f()" />:
+
+---
+
+<Math code="%5Clbrace%20pre%5C_f%20%5Crbrace%20%5C%3B%5C%3B%20f()%20%5C%3B%5C%3B%20%5Clbrace%20post%5C_f%20%5Crbrace" />
```

Comment:
The subscript "_f" doesn't render as subscript in this version.

Reply:
Fixed